Enhancement proposal - CNF-tests split

liornoy opened the pull request:
This commit adds a new proposal for spliting the cnf-tests Including re-write of the cnf-tests container making it contain only the latency tests.

zeeke commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,217 @@
+In order to maintain the same reporting capabilities we need to:
+ 1. Add the k8s reporter to each repository that doesn't has it already.
+ 2. Use ginkgo's built-in junit generator, so that we don't have to add the `-junit` flag to each repo.
+ 3. Create a local junit-merger program to aggregate all of the junit reports under the same test suite.
```
Is there any specific requirement for merging all the unit reports into one? Prow will collect any `junit*.xml` , so it can be fine to have separated ones

liornoy replied:
It's just what is necessary for us to keep the same 3 junit report files: one for each test suite
We even made recently a soft link to not break backward compatibility:
```
ln -sf "$TESTS_REPORTS_PATH/junit_setup.xml" "$TESTS_REPORTS_PATH/setup_junit.xml"
ln -sf "$TESTS_REPORTS_PATH/junit_cnftests.xml" "$TESTS_REPORTS_PATH/cnftests-junit.xml"
ln -sf "$TESTS_REPORTS_PATH/junit_validation.xml" "$TESTS_REPORTS_PATH/validation_junit.xml"
```

cgoncalves commented:
@sshnaidm FYI while the API is expected to remain unchanged, there will be changes that may impact CI.

sshnaidm commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+   we will have to include "/raw" subpath, where we create all the junit reports, 
+   and a "/junit" subpath where we hold the merged junit reports.
+
+### Rewrite the cnf-tests container
```
Just a clarification, we don't talk here about `quay.io/openshift-kni/cns-tests:..` container, that is used in performance tests? Or do we?

liornoy replied:
Yes, it's that.

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+### Workflow Description
+
+This change will not affect the current API, including parameters/environment variables used to invoke the cnf-tests: `TEST_SUITES`, `FEATURES`, `FOCUS_TESTS`, `SKIP_TESTS`, `GINKGO_PARAMS`, `FAIL_FAST`, and `TESTS_REPORTS_PATH`. 
+However, with this proposal, instead of running test binaries that combine all the CNF components,  we will introduce an **optional** parameter named `CNF_COMPONENTS` to control what external test suites to run.  
```
The CNF_COMPONENTS wording is misleading. If this selects what external tests to run, let's call it "EXTERNAL_SUITES" or something like that

liornoy replied:
Agreed. I'll change it, thanks.

fedepaol commented on enhancements/cnf-tests-split.md:
````diff
@@ -0,0 +1,204 @@
+
+The mapping of suite + component to the external test suite path is done by:
+```bash
+declare -A TESTS_PATHS=\
````
Can we go with convention over configuration and expect the extenral suite to match the name of the local path of a given suite?

liornoy replied:
Do you mean that the user will set EXTERNAL_SUITE to the local path e.g. `"cnf-tests/submodules/metallb-operator/test/e2e/validation"` instead of just metallb?
What are the benefits of going with the approach?

fedepaol replied:
I was thinking more something like "validation/metallb" (without the cnf-tests because it's implicit, without submodules because I am not sure it makes sense, see my comment below).
The advantage is not needing to maintain the map if we add / remove a suite.

liornoy replied:
resolved the conversation in the comment below

fedepaol commented on enhancements/cnf-tests-split.md:
````diff
@@ -0,0 +1,204 @@
+```bash
+declare -A TESTS_PATHS=\
+(["configsuite nto"]="cnf-tests/submodules/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/0_config"\
+ ["validationsuite cluster"]="cnf-tests/testsuites/validationsuite"
````
this is more "integration" rather than "cluster"

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+ ["validationsuite cluster"]="cnf-tests/testsuites/validationsuite"
+ ["validationsuite metallb"]="cnf-tests/submodules/metallb-operator/test/e2e/validation"
+ ["cnftests integration"]="cnf-tests/testsuites/e2esuite"\
+ ["cnftests metallb"]="cnf-tests/submodules/metallb-operator/test/e2e/functional"\
```
Can we have paths like

cnf-tests/external/validation/metallb-operator
cnf-tests/external/validation/sriov

and 

cnf-tests/external/e2e/metallb-operator
cnf-tests/external/e2e/sriov

or something like that to make the navigability easier?

liornoy replied:
This map needs to point to the full path of the test suite in order to invoke it later in the script.
Do you suggest creating soft links to those paths while still cloning to the submodule dir?

fedepaol replied:
is the name of the submodule mandatory to be "submodule" ? do we really care if a test suite that we find in a local path comes from a submodule or not?

liornoy replied:
can you please elaborate on how will this work when a test suite in a local path comes not from a submodule?

fedepaol replied:
Let's remove the external part then.

cnf-tests/validation/metallb-operator
cnf-tests/validation/sriov
cnf-tests/validation/integration

and

cnf-tests/e2e/metallb-operator
cnf-tests/e2e/sriov
cnf-tests/e2e/integration

Is it possible? And in general, can we think about a dir structure where it's clear what's where without needing the map?

fedepaol replied:
ah now I understand. Well, it's not optimal and it makes the navigability of hte project a bit harder. On the other hand I am not sure reshuffling the folders of a submodule makes sense, nor having symlinks.

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+
+To invoke the cnf-tests we'll use the same `Makefile` targets as before. e.g.: `functests-on-ci`, `feature-deploy-on-ci`, etc.
+Those rules will have a prerequisite to run the new `init-git-submodules` target which inits and updates the git submodules.
+The `init-git-submodules` will include the following env vars that control the commit hash of the components:
```
Can we enforce that those submodules were inited and that their commit corresponds to what we have somewhere? It'd make sense to have a single source of truth for those hashes. This would prevent false negatives or us runnign the wrong set of tests.

fedepaol replied:
Yep, something like that

liornoy replied:
Since we've decided on pointing to the latest commit, this conversation about validating we inited to the correct commit is outdated.

fedepaol commented on enhancements/cnf-tests-split.md:
````diff
@@ -0,0 +1,204 @@
+Those rules will have a prerequisite to run the new `init-git-submodules` target which inits and updates the git submodules.
+The `init-git-submodules` will include the following env vars that control the commit hash of the components:
+```bash
+export METALLB_OPERATOR_TARGET_COMMIT?= main
````
As I wrote above, pinning to a given branch is a bad idea imo

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+will remove/change this flag, it may cause the test run to fail,
+and it is out of our control.  
+We might break some internal users' pipelines in case they didn't migrate
+and kept building and using the cnf-tests container for running the cnf-tests.
```
Given that upstream (openshift-kni/cnf-features-deploy) will be the source of truth for the pinned dependencies, we can mitigate this by always run all the tests with dry-run. That would tell us if some suite broke with regards to our "api" before committing.

liornoy replied:
Good idea. just running a sanity check before the real tests. and if it fails it would be clear what's the cause of it.
i.e. someone removed/change the `-reporter` flag

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+
+In order to implement this we will have to touch other code bases and
+add the k8s reporter, which can be rejected/unwelcomed by the upstream community.  
+Also, the k8s reporter flag we introduce in the external repositories 
```
This is only if  we invoke with the reporter flag, right?

liornoy replied:
right. which we currently do.

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+
+# CNF-tests split
+
+The cnf-tests is a sub-project in the cnf-features-deploy repository, aimed to run tests against various CNF workloads on top of an Openshift cluster.  
```
This is not exactly true. The cnf tests suite aims to validate the various openshift features necessary to run cnf workloads on a cluster.

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+
+## Summary
+
+The purpose of this enhancement is to decouple the cnf-tests from the test suites of the external CNF projects. Additionally, as this change affects the way tests are run, we propose rewriting the cnf-tests container image, which is a supported product for running the latency tests.
```
of the features hosted in external projects. The projects are not "CNF"

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+
+## Proposal
+### User Stories
+#### Story 1
```
These stories don't tell what is gonna change after this. I feel we should focus on the bumping / updating process rather on something we are not planning to change (which is how we run the tests and what tests we run today).
We can add a global item "the way we run the tests / what tests we run won't change" to collect all of this.

liornoy replied:
added the new item.

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+The most major risk here is revolving around changing the cnf-tests container.  
+That is because it's delivered as a product supported by Red Hat, and
+we should make sure to keep backward compatibility and provide the relevant documentation, explaining the new interface to the latency tests.  
+We should also provide a mitigation guide for the various internal users 
```
What mitigations we expect to need if we maintain the same API? If we expect any, can we list them?

liornoy replied:
my mistake, the api is the same so there's nothing different the users should do

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+
+## Design Details
+### Test Plan
+1. Validate that the cnf-tests are passing on CI.
```
Not only that they are passing. We should ensure that the set of tests we expect to run are effectively running

liornoy replied:
added

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+
+### Workflow Description
+
+This change will not affect the current API, including parameters/environment variables used to invoke the cnf-tests: `TEST_SUITES`, `FEATURES`, `FOCUS_TESTS`, `SKIP_TESTS`, `GINKGO_PARAMS`, `FAIL_FAST`, and `TESTS_REPORTS_PATH`. 
```
how about discovery mode?

fedepaol replied:
And how about the deployment mechanism that currently leverages some of the same env variables?

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+To explain this new variable further, read the following example:
+Running the `e2esuite` test suite, with an empty `CNF_COMPONENTS` and `FEATURES=metallb`, will result in 
+running ginkgo's test in all of the components: sriov, metallb, nto-performance, nto-latency, ptp and integration, adding the flag `-ginkgo.focus=metallb`.
+So in other words, the functionallity of the `FEATURES` var stays the same.
```
typo: functionallity

fedepaol commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+1. Stop vendoring the external test suites and run the tests directly from the projects.
+2. Maintain the same API for running the tests.
+3. Preserve the same testing report output.
+4. Refactor the cnf-tests container to contain only the latency-tests.
```
I'd enforce the concept and say that the only way to run cnf tests will be by cloning the repo.

yuvalk commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+
+To explain this new variable further, read the following example:
+Running the `e2esuite` test suite, with an empty `CNF_COMPONENTS` and `FEATURES=metallb`, will result in 
+running ginkgo's test in all of the components: sriov, metallb, nto-performance, nto-latency, ptp and integration, adding the flag `-ginkgo.focus=metallb`.
```
we used to have a weekend latency run (because to have significant, comparable results, it needs to run for ~10hours

liornoy replied:
The `nto-latency` I mention here is part of the e2esuite: [link](https://github.com/openshift-kni/cnf-features-deploy/blob/ce9c1d113fd1d6e35db37270156744230c6fdd14/cnf-tests/testsuites/e2esuite/test_suite_test.go#LL38C13-L38C71)
It is not the classic latency run that requires setting different parameters. see more in the rewrite cnf-tests container section.

liornoy commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+In order to maintain the same reporting capabilities we need to:
+ 1. Add the k8s reporter to each repository that doesn't has it already.
+ 2. Use ginkgo's built-in junit generator, so that we don't have to add the `-junit` flag to each repo.
+ 3. Create a local junit-merger program to aggregate all of the junit reports under the same test suite.
```
The `junitparser` is a program you need to download from `pip` don't we prefer not being dependent on that?

sshnaidm replied:
Why not? You can write your own merger of course and test it and fix bugs there, but not sure worth to reinvent the wheel. Up to you of course.

liornoy replied:
Mostly because I don't want to introduce a new dependency for the cnf-tests users, and have this change as frictionless as possible, providing the same report files as before.

liornoy commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+ 1. Add the k8s reporter to each repository that doesn't has it already.
+ 2. Use ginkgo's built-in junit generator, so that we don't have to add the `-junit` flag to each repo.
+ 3. Create a local junit-merger program to aggregate all of the junit reports under the same test suite.
+ 4. Change the reports output directory. Having `TESTS_REPORTS_PATH` as the root path,
```
yes

liornoy commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+ 2. Use ginkgo's built-in junit generator, so that we don't have to add the `-junit` flag to each repo.
+ 3. Create a local junit-merger program to aggregate all of the junit reports under the same test suite.
+ 4. Change the reports output directory. Having `TESTS_REPORTS_PATH` as the root path,
+   we will have to include "/raw" subpath, where we create all the junit reports, 
```
ok, we'll discuss it offline. but anyway, to clarify, the additional `/raw` and `/junit` subdirectories are required to aggregate the reports of every test suite together. That is because, with the proposal, every suite includes a number of CNF components that each one produces a JUnit report. and after we run a test suite we want to merge all of the reports.

sshnaidm replied:
I understand, it kind of what we did when ran in parallel a few ginkgo processes. Anyway, creating directories in CI shared directories is problematic, so I need an option to use it "flat".

liornoy replied:
I'll update you on this, that this section was removed and we'll produce the reports without additional directories.

liornoy commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+## Summary
+
+The purpose of this enhancement is to decouple the cnf-tests from the test suites of the external CNF projects. Additionally, as this change affects the way tests are run, we propose rewriting the cnf-tests container image, which is a supported product for running the latency tests.
+## Motivation
```
It is running against a linter as part of the CI. apparently this wasn't caught

liornoy commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+2. Some dependencies will still be required in the components as there are tests that combine multiple features.
+
+
+This change will also allow us to pin the external repositories to the latest branch, ensuring that we always test the latest tests without the need for manual commit hash updates. We will include an option to set an environment variable that specifies the commit ref for each repo.
```
I remember from the meetings we had that we did want to always test the latest, and being pinned to a specific commit was a disadvantage.
You are right that it will cause failures to show up more frequently (ci fails more frequently), but isn't it the point?
to catch it early and fix it before shipping it to the customers?
I would like to hear more opinions about that please: @cgoncalves @yanirq @sshnaidm

fedepaol replied:
I don't remember this discussion, and not having control of what is running where is a recipe for disaster. Executions must be consistent as much as we can.
It's not about the frequency of the failures, is about knowing what tests are running and being able to reproduce.

sshnaidm replied:
I'm not sure I understand the point of manual pinning. What is the criteria for update then?
If it's passing all tests, then you anyway should run CI jobs on it.

sshnaidm replied:
Well, somebody should pass CI ideally for committing the test, and that supposes to block breakages. But I see your point, it can be a flaky test for example. I'm not against having a pinning mechanism for returning to a passing point, but not sure it will be updating all the time. People usually submit test and go. We have a real danger to run old and probably not relevant tests because nobody updates it here.
But it's always good to have option to pin it to a previous version where tests were passing until it's fixed.

liornoy replied:
Maybe by combining both approaches, we will both pin to a specific commit to be on the safe side, plus
create an automated bump script like we have in metallb-operator and thus we could catch if something is failing in the PR's ci.
This script will save us time updating all the time and will create the bump PR for us automatically.

fedepaol replied:
Let me try to expand. With the current version, we are in control of what test ran in what day (because we bump the tests manually). This allows us to have a reproducible environment when a given test fails, which allows us to understand if the issue was caused by a test or by the feature that we want to test.

On top of that, given we gate (at least some features) with CI, bumping the tests with a PR will allow us to catch those issues early, because the bump PR itself will fail, giving a way more narrowed scope compared to the periodic CI. This is better, imo, to having to triage the periodic CI that starts to fail.

This requires the same effort that we put in today, by aligning each feature, with the advantage that we can automate the mechanism as @liornoy said, because now it's about bumping a commit instead of having to deal with go dependencies.

I'd also like to hear @SchSeba 's view here

sshnaidm replied:
I understand the need of stable CI in this repo or wherever. For Upstream CI it's not the case, we need to latest code to use for tests, usually this code contains latest fixes as well. It doesn't make sense to run CI on the stable code, its purpose to run on the most fresh in order to catch issues ASAP and as close to the code as possible. Running on old code is something else, not CI.
Although it might work with automated PRs, but I very doubt it, probably it will include both fixes for old issues and new bugs, so it might be blocked for updates for a long time, also someone will need to maintain it, which is might be a problem by itself in this repo that always needed more love.
Anyway, I don't mind to have any updating mechanisms here, but for our CI we must have the latest code, if it fails - then bugs should be submitted and resolved (and probably tests skipped until resolved). If it will be submodules, I probably can clone the repo and force submodules to fetch the most fresh code instead of pinned one. If it will be done in a different way, then we'll need other ways to have the newest code. But this is a **hard requirement** from our side.

fedepaol replied:
> I understand the need of stable CI in this repo or wherever. For Upstream CI it's not the case, we need to latest code to use for tests, usually this code contains latest fixes as well. It doesn't make sense to run CI on the stable code, its purpose to run on the most fresh in order to catch issues ASAP and as close to the code as possible. 

I agree. What I am advocating here is not because of stability, it is because of consistency and reproducibility
Running against "main" means "we don't know what version of tests we are running""

> Running on old code is something else, not CI. Although it might work with automated PRs, but I very doubt it, probably it will include both fixes for old issues and new bugs, so it might be blocked for updates for a long time, also someone will need to maintain it, which is might be a problem by itself in this repo that always needed more love

This is what I was saying above. If a PR that bumps the tests in a way that makes them fail, it's better to know from the PR and not from the failing CI, because it's clearer to know what is causing the failure. Not pinning means hiding this under the carpet and making it harder to find that a change in the tests broke them.

>  Anyway, I don't mind to have any updating mechanisms here, but for our CI we must have the latest code, if it fails - then bugs should be submitted and resolved (and probably tests skipped until resolved). If it will be submodules, I probably can clone the repo and force submodules to fetch the most fresh code instead of pinned one. If it will be done in a different way, then we'll need other ways to have the newest code. But this is a **hard requirement** from our side.


If downstream CI will override the pinning then there's no point in the discussion. It will just make triaging failures  harder for the reasons I stated above.

sshnaidm replied:
> I agree. What I am advocating here is not because of stability, it is because of consistency and reproducibility Running against "main" means "we don't know what version of tests we are running""
> 

I'm not sure why it's not possible to know what version of tests we are running. If it's a technical problem, it can be solved easily saving commit hashes somewhere. Jenkins does it automatically, and in Prow we'll need probably a primitive script to do it. Later by looking between hashes of success and failure we can see where the problem is. Isn't it as usually do?
As I understood the main problem for you is that while investigating and trying to fix the bug, the CI will continue to fail. Correct me if I'm wrong.
In our case we're implementing skipping of failing tests.

> 
> If downstream CI will override the pinning then there's no point in the discussion. It will just make triaging failures harder for the reasons I stated above.

I don't know if they do, but I discovered recently downstream CI uses `nto` repository from upstream for example, and that provides them fixes faster than we have in "upstream CI", because we're stuck here with old versions... That's weird.
cc @yanirq

liornoy replied:
I just spotted an issue, and I don't know if that's a deal breaker, and it's just some technical thing, but if we agreed to keep the cnf-tests docs (DOCGEN stuff) then it must be updated via PR every time a new test is added/removed/updated.
Pinning to the main branch will cause the `make validate-test-list` to fail without properly editing the `TESTLIST.md`. correct me if I'm wrong.

fedepaol replied:
> As I understood the main problem for you is that while investigating and trying to fix the bug, the CI will continue to fail. Correct me if I'm wrong.

Nope, if there's a failure and there's a bug related to that failure, I don't mind if that test keeps failing or not, because we are aware there's an issue to be triaged.

My main problem is not being in control of what runs (and what ran, at a given point in time), and about swallow whatever change is applied to the tests before validating them, and getting a "all tests failed" report just because the location of the e2e tests moved in some repo we depend on.

The "what ran" can be solved by dumping the commit hashes as you said, the second part probably not, but we'll live with that I guess. 
Note that this also change the semantics of this repo, which becomes just glue scripts for running single test suites in an uniform matter, rather than a stand alone entity that can be consumed at any given point in time, because every time we run it, it may run a different set of tests depending on what "latest" contains.

That goes together with @liornoy point of not being able to document the tests contained by the suite.

Said that, I don't want to hold the design on this item. As long as we are able to know what version of the tests ran and what version of each feature was deployed at a given time, I am good.

liornoy commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+### User Stories
+#### Story 1
+
+As a developer, I want to run the cnf-tests suites so that I can validate my CNF feature works properly.
```
I'll add this. thanks.

liornoy commented on enhancements/cnf-tests-split.md:
````diff
@@ -0,0 +1,204 @@
+)
+```
+
+To prevent conflicts due to the numerous dependencies in the test suites, we want to decouple the external test suites by introducing a script that invokes the required test suites as git-submodules under the cnf-tests/submodules path.
````
I'll reword this. thanks.

liornoy commented on enhancements/cnf-tests-split.md:
```diff
@@ -0,0 +1,204 @@
+
+### Goals
+
+1. Stop vendoring the external test suites and run the tests directly from the projects.
```
I'll reword this. thanks.

liornoy replied:
Just to make sure we're on the same page: the git submodule doesn't just clone the test portion, but the whole repo itself.

yanirq commented on enhancements/cnf-tests-split.md:
````diff
@@ -0,0 +1,237 @@
+)
+```
+
+To prevent conflicts due to the numerous dependencies in the test suites, we want to decouple the external test suites by cloning them indipendently via git-submodules and have a script that just invokes them, according to the given environment variables.
````
s/indipendently/independently

liornoy replied:
thanks. fixed

yanirq commented:
overall the proposal looks good, I would like @shajmakh to take a look as well with emphasis on the refactor for using cnf-tests image only for latencty tests.

liornoy commented:
Updates to the proposal:
* Fixed typos.
* Made the git submodules init to the latest commit by default.
* Added a comment about the test list documentation.
* Added a section about migration actions with an example.

shajmakh commented:
@liornoy Looks good to me.

sshnaidm commented:
@liornoy thanks, lgtm

evgenLevin commented:
/lgtm

SchSeba reviewed (commented):
/lgtm

SchSeba commented:
/approve

ffromani commented:
/approve

on behalf of other reviewers, trusting them
